### ISSUE TITLE
Deeplab does not produce correct segmentation mask.

### ISSUE THREAD

**Issue by ajinkya933** [tensorflow/models#6565]
------------------------

### System information
- **What is the top-level directory of the model you are using**: models->research ->Deeplab
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 16.04
- **TensorFlow installed from (source or binary)**: source
- **TensorFlow version (use command below)**:1.13
- **Bazel version (if compiling from source)**: bazel-0.24.0-installer-darwin-x86_64

### Describe the problem

I am training my own model to segment squares in images using Deeplab. For that I have annotated the training data, Heres an example of training image(on left) and segmentation mask  image(on right).

<img width="1042" alt="Screenshot 2019-04-12 at 12 29 17 PM" src="https://user-images.githubusercontent.com/17012391/56018029-9911e680-5d1e-11e9-9d51-74b839f58b7a.png">

I am converting this training data to record file using: Deeplab-> build_voc2012_data.py

I have further made following changes to the scripts described as below, These changes are such that I can train my own dataset.

**deeplab -> datasets -> datagenerator.py**

> TESTSET_INFORMATION = DatasetDescriptor(
>     splits_to_sizes={
>         'train': 800,  # num of samples in images/training
>         'val': 200,  # num of samples in images/validation
>     },
>     num_classes=2,
>     ignore_label=255,    # When setting ignore_label = 0 predictions come back as a red or green image; and when setting ignore_label = 255 predictions come back as a black image
> )
> _DATASETS_INFORMATION = {
>     'cityscapes': _CITYSCAPES_INFORMATION,
>     'pascal_voc_seg': _PASCAL_VOC_SEG_INFORMATION,
>     'ade20k': _ADE20K_INFORMATION,
>     'testset': _TESTSET_INFORMATION,
> }
> 

**utils > train_utils.py**

> exclude_list = ['global_step', 'logits'] 

> not_ignore_mask = tf.to_float(tf.equal(scaled_labels, 0)) * 1 + tf.to_float(tf    .equal(scaled_labels, 1)) * 2 + tf.to_float(tf.equal(scaled_labels, ignore_label    )) * 0

Please note batch size=12. And training number of steps = 1000. I get following predictions

<img width="1088" alt="Screenshot 2019-04-12 at 12 37 56 PM" src="https://user-images.githubusercontent.com/17012391/56018457-cf039a80-5d1f-11e9-96bf-5cf6171619b1.png">

I am not sure what is it I am doing wrong ?
To further debug the problem I observed that during training the loss dosen't change it keeps on oscillating.

example loss is:

> Total loss is :[0.0697127208]
> INFO:tensorflow:global_step/sec: 0.715997
> Total loss is :[0.0697126836]
> INFO:tensorflow:global_step/sec: 0.835336
> Total loss is :[0.0697126538]
> INFO:tensorflow:global_step/sec: 0.83706
> Total loss is :[0.0697126389]
> INFO:tensorflow:global_step/sec: 0.833028
> Total loss is :[0.069712624]
> INFO:tensorflow:global_step/sec: 0.836747
> Total loss is :[0.069712624]
> INFO:tensorflow:global_step/sec: 0.831312
> Total loss is :[0.069712624]
> INFO:tensorflow:global_step/sec: 0.832811
> Total loss is :[0.069712624]
> INFO:tensorflow:global_step/sec: 0.83014
> Total loss is :[0.069712624]
> INFO:tensorflow:global_step/sec: 0.830697
> 
> 

**What should I do so that the loss is reduced and the model gives a proper square as a segmentation output ?**

@bleedingfight @aquariusjay  @Citygity could you shed some light and help me bring one step closer to the solution ?

**Comment by ajinkya933** [comment]
I went through issue #3739 and I am referring to zhaolewen's solution:

After quite some struggles, I finally got the deeplab model running on a single fore-ground object class segmentation task. I find the configuration of the label images very important:

1) num_classes = num of fore-ground object class count + background, thus in my case it's 2

2)ignore_label=255 means that in the single-channel png label image, the color 255 will mark the region that does not influence the calculation of loss and gradients. (thus, pay attention not to mark your object as 255)

3) the value of the class of the object is the exact value in the label png image. that is to say, if you have 5 classes, they should be marked as 0,1,2,3,4 in the png. They will not be visible in the png but that's ok. Do not give them values like 40, 80, 120, 180, 220 to make them visible, because the code in the repo reads the exact values in the png as the label of the class.

4) in the label png, 0 is the background. no not set ignore_label=0 (to not complicate things)

I am doing everything correct except point 3 I do not understand point 3.

**Comment by vijay1131** [comment]
@ajinkya933 How did you solved your issue?

**Comment by essalahsouad** [comment]
@ajinkya933 did you solved the problem??

**Comment by ajinkya933** [comment]
My input images which I used for training were not used properly, 
Make sure that your object has pixel value of RGB(1,1,1) and your background has pixel value(0,0,0)

**Comment by essalahsouad** [comment]
@ok, thank you
